File: Auto/final_code/c-c++/omnibot.cpp
#include "MotorModule.h"
#include "LaneModule.h"
#include "WebcamModule.h"

int main() {
    // Initialisiere die Motorsteuerung mit den entsprechenden Pins
    Motor motor(10, 9, 11, 17, 22, 27);

    // Erfasse ein Bild von der Webcam
    cv::Mat img = WebcamModule::getImg();

    // Berechne die Kurvenbewertung für die Fahrspurerkennung
    double curveVal = getLaneCurve(img, 1);

    // Empfindlichkeit und maximale Geschwindigkeit einstellen
    double sen = 1.3;  // Empfindlichkeit
    double maxVal = 0.3; // Maximale Geschwindigkeit

    // Beschränke die Kurvenbewertung auf die maximale Geschwindigkeit
    if (curveVal > maxVal)
        curveVal = maxVal;
    if (curveVal < -maxVal)
        curveVal = -maxVal;

    // Feine Anpassungen der Empfindlichkeit je nach Kurvenbewertung vornehmen
    if (curveVal > 0) {
        sen = 1.7;
        if (curveVal < 0.05)
            curveVal = 0;
    } else {
        if (curveVal > -0.08)
            curveVal = 0;
    }

    // Motor mit Geschwindigkeit und Kurvenbewertung steuern
    motor.move(0.20, -curveVal * sen, 0.05);

    return 0;
}

File: Auto/final_code/python/omnibot.py
# Importieren der erforderlichen Module
from MotorModule import Motor
from LaneModule import getLaneCurve
import WebcamModule
import time
import psutil
import os

# Setzen der Umgebungsvariable für das SDL-Video auf "dummy" (headless-Modus)
os.environ["SDL_VIDEODRIVER"] = "dummy"

# Initialisierung des Motors mit den entsprechenden GPIO-Pins
motor = Motor(10, 11, 9, 17, 22, 27)

# Aktuelle Prozess-ID und CPU-Auslastung abrufen
pid = os.getpid()
py = psutil.Process(pid)
cpu_use = py.cpu_percent(interval=1)

# Hauptfunktion
def main():
    start_time = time.time()
    
    # Speichernutzung in GB anzeigen
    memoryUse = py.memory_info()[0] / 2.0**30  # Speichernutzung in GB...
    print('Speichernutzung:', memoryUse, 'GB')
    
    # Einzelbild von der Webcam erfassen
    img = WebcamModule.getImg()
    
    # Fahrspurkurve bestimmen
    curveVal = getLaneCurve(img, 0)
    
    # Maximalgeschwindigkeit und Kurvenempfindlichkeit festlegen
    maxVal = 0.03  # Maximale Geschwindigkeit
    if curveVal > maxVal:
        curveVal = maxVal
    if curveVal < -maxVal:
        curveVal = -maxVal
    
    # Motorbewegung basierend auf der Kurve ausführen
    motor.move(0.06, curveVal, 0.01)
    
    # CPU-Auslastung anzeigen
    print("CPU-Auslastung: " + str(psutil.cpu_percent()) + "%")
    
    end_time = time.time()
    execution_time = end_time - start_time
    print(f'Ausführungszeit: {execution_time} Sekunden')

if __name__ == '__main__':
    while True:
        main()
